core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -528,7 +528,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       }
 
     // Optionally scale number of executors dynamically based on workload. Exposed for testing.
-    val dynamicAllocationEnabled = _conf.getBoolean("spark.dynamicAllocation.enabled", false)
+    val dynamicAllocationEnabled = _conf.getBoolean("spark.dynamicAllocation.enabled", false) &&
+      (_conf.getInt("spark.executor.instances", 0) == 0)
```

Comment:
yes, +1 to @srowen's comment. We also need to add a comment here to explain why it's not enabled if we set `spark.executor.instances`, something like

```
// Dynamic allocation and explicitly setting the number of executors are inherently
// incompatible. In environments where dynamic allocation is turned on by default,
// the latter should override the former (SPARK-9092).
```